modules/dop/dao/issue.go

```diff
@@ -620,3 +643,64 @@ func (client *DBClient) GetIssueNumByPros(projectIDS []uint64, req apistructs.Is
 
 	return res, nil
 }
+
+var joinState = "LEFT JOIN dice_issue_state ON dice_issues.state = dice_issue_state.id"
+
+type IssueExpiryStatus struct {
+	IssueNum     uint64
+	ProjectID    uint64
+	ExpiryStatus ExpireType
+}
+
+func (client *DBClient) GetIssueExpiryStatusByProjects(req apistructs.WorkbenchRequest) ([]IssueExpiryStatus, error) {
+	sql := client.Table("dice_issues").Joins(joinState).Select("count(dice_issues.id) as issue_num, dice_issues.project_id, dice_issues.expiry_status")
+	sql = sql.Where("assignee = ? AND dice_issue_state.belong IN (?)", req.Assignees, req.StateBelongs)
+	if len(req.ProjectIDs) > 0 {
+		sql = sql.Where("dice_issues.project_id IN (?)", req.ProjectIDs)
+	}
+	offset := (req.PageNo - 1) * req.PageSize
+	var res []IssueExpiryStatus
+	if err := sql.Offset(offset).Limit(req.PageSize).Group("dice_issues.project_id, dice_issues.expiry_status").Find(&res).Error; err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+func (client *DBClient) GetIssuesByProject(req apistructs.IssuePagingRequest) ([]Issue, uint64, error) {
+	var res []Issue
+	sql := client.Table("dice_issues").Joins(joinState)
+	sql = sql.Where("dice_issues.project_id = ? AND assignee = ? AND dice_issue_state.belong IN (?)", req.ProjectID, req.Assignees, req.StateBelongs)
+	if req.OrderBy != "" {
+		if req.Asc {
+			sql = sql.Order(fmt.Sprintf("%s", req.OrderBy))
+		} else {
+			sql = sql.Order(fmt.Sprintf("%s DESC", req.OrderBy))
+		}
+	}
+	var total uint64
+	offset := (req.PageNo - 1) * req.PageSize
+	if err := sql.Offset(offset).Limit(req.PageSize).Find(&res).Offset(0).Limit(-1).Count(&total).Error; err != nil {
+		return nil, 0, err
+	}
+	return res, total, nil
+}
+
+func (client *DBClient) BatchUpdateIssueExpiryStatus(states []apistructs.IssueStateBelong) error {
+	sql := `
+	UPDATE dice_issues a LEFT JOIN dice_issue_state b ON a.state = b.id
+	SET a.expiry_status = 
+	(
+	CASE
+	  WHEN a.plan_finished_at IS NULL THEN 'Undefined'
+	  WHEN a.plan_finished_at < CURDATE() THEN 'Expired'
+	  WHEN a.plan_finished_at = CURDATE() THEN 'ExpireIn1Day'
+	  WHEN a.plan_finished_at = DATE_ADD(CURDATE(),INTERVAL 1 DAY) THEN 'ExpireIn2Day'
+	  WHEN a.plan_finished_at < DATE_ADD(CURDATE(),INTERVAL 7 DAY) THEN 'ExpireIn7Day'
+	  WHEN a.plan_finished_at < DATE_ADD(CURDATE(),INTERVAL 30 DAY) THEN 'ExpireIn30Day'
+	  ELSE 'ExpireInFuture'
+	END
+	)
+    WHERE a.plan_finished_at IS NOT NULL AND b.belong IN (?)
+	`
```

Comment:
I think you should split to 7 sql, because your current sql may update full-table.

```sql
UPDATE dice_issues a LEFT JOIN dice_issue_state b ON a.state = b.id
SET a.expiry_status = 'Expired'
WHERE a.plan_finished_at < CURDATE() AND a.expiry_status != 'Expired'
```
and so on, the key is **a.expiry_status != 'Expired'**.

Reply:
After we split sql and add index. For single sql with one million data, time reduced from 13s to 1.7s